File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const bfs = (start, end) => {
  const queue = [];
  queue.push(start);
  visited[start] = 0;

  while (queue.length > 0) {
    const current_idx = queue.shift();
    if (current_idx === end) {
      return visited[end];
    }
    for (let i = 1; i <= 6; i++) {
      let next_idx = current_idx + i;
      if (next_idx <= 100 && visited[next_idx] == 0) {
        if (t[next_idx]) {
          next_idx = t[next_idx]
        }
        if (visited[next_idx] === 0) {
          queue.push(next_idx);
          visited[next_idx] = visited[current_idx] + 1;
        }
      }
    }
  }
};

// 100번 칸을 넘어간다면 이동할 수 없다.
const [n, m] = input[0].split(" ").map(Number); // 게임 판에 있는 사다리의 수 n, 뱀의 수 m
const t = {};
for (let i = 1; i <= n + m; i++) {
  const [s, e] = input[i].split(" ").map(Number);
  t[s] = e;
}
const visited = Array(101).fill(0);

let start = 1;
const end = 100;
console.log(bfs(start, end));

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
# 뱀과 사다리 게임

# 게임의 크기는 10 * 10
# 사다리의 수 n, 뱀의 수  m
# n개의 줄에 사다리의 정보를 의미하는 x, y
# m개의 줄에 뱀의 정보를 의미하는 u, v

import sys
input = sys.stdin.readline
from collections import deque



def bfs(start,end):
    queue = deque()
    queue.append([start,0])
    while queue:
        now, value = queue.popleft()
        visited[now] = value

        if now == end :
            return visited[now]

        for i in range(1,7):
            nxt_now = now + i
            if nxt_now < 101 and visited[nxt_now] == 0:
                if nxt_now in ladders.keys():
                    nxt_now = ladders[nxt_now]
                visited[nxt_now] = value + 1
                queue.append([nxt_now, value+1])

n, m = map(int,input().split())
ladders = dict()
visited = [0 for _ in range(101)]
for _ in range(n):
    x, y = map(int,input().split())
    ladders[x] = y

for _ in range(m):
    u, v = map(int,input().split())
    ladders[u] = v


print(bfs(1,100))